src/Features/Core/Portable/Completion/Providers/Snippets/AbstractSnippetCompletionProvider.cs

```diff
@@ -0,0 +1,65 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.Completion.Providers.Snippets
+{
+    internal abstract class AbstractSnippetCompletionProvider : CommonCompletionProvider
+    {
+        public abstract Task<SyntaxToken> GetTokenFromDocumentAsync(Document document, int position, CancellationToken cancellationToken);
+
+        public AbstractSnippetCompletionProvider()
+        {
+
+        }
+
+        public override async Task<CompletionChange> GetChangeAsync(Document document, CompletionItem item, char? commitKey = null, CancellationToken cancellationToken = default)
+        {
+            var service = document.GetRequiredLanguageService<ISnippetService>();
+            var tokenSpanStart = SnippetCompletionItem.GetTokenSpanStart(item);
+            var tokenSpanEnd = SnippetCompletionItem.GetTokenSpanEnd(item);
+            var span = item.Span;
+
+            var snippetProvider = service.GetSnippetProvider(new SnippetData(item.DisplayText));
```

Comment:
is there a better way to do this to get the provider I want?

Reply:
Not saying this is a better way, but this is how it works for completion item/provider
https://github.com/dotnet/roslyn/blob/main/src/Features/Core/Portable/Completion/CompletionServiceWithProviders.ProviderManager.cs#L95

Reply:
> using the display text seems dangerous (esp. if for loc reasons or importing reasons, we get multiple with teh same name). can you give them all unique ids which get added as properties of hte completion item. that way you can always get back uniquely and safely.

yeah, that would make sense.